### ISSUE TITLE
Better interface names

### ISSUE THREAD

**Issue by schnittstabil** [http-message-strategies-interop/fig-standards#3]
As objected by @mindplay-dk at  https://github.com/http-message-strategies-interop/fig-standards/issues/2#issuecomment-281479642, the interface names are by far too meaningless :cry:

In case of `__invoke` interfaces, the interface name must solely describe what it does – not the method name; hence, better names are needed.

The current interfaces can always be found at [proposed/http-message-strategies/http-message-strategies.md](https://github.com/http-message-strategies-interop/fig-standards/blob/http-message-strategies/proposed/http-message-strategies/http-message-strategies.md#1-specification).

**Comment by schnittstabil** [comment]
**Note** `http-message-strategies` branch updated, but the names are still open for debate.

Current interface names at `Psr\Http\Message\Strategies`:

```
       RequestHandlerInterface :       RequestInterface → ResponseInterface
 ServerRequestHandlerInterface : ServerRequestInterface → ResponseInterface

      RequestOperatorInterface :       RequestInterface → RequestInterface
ServerRequestOperatorInterface : ServerRequestInterface → ServerRequestInterface
     ResponseOperatorInterface :      ResponseInterface → ResponseInterface
```

**Comment by mindplay-dk** [comment]
@schnittstabil I'm glad to see you took this to heart :-)

Will watch this thread to see where it goes from here. thanks!

**Comment by schnittstabil** [comment]
The more I think about it, the more I believe the most accurate name for `__invoke` interfaces include a `Function` suffix, e.g.:

```php
interface FilterFunction
{
    public function __invoke($value); 
}

class IsEvenFilter implements FilterFunction
{ 
    public function __invoke($value) 
    { 
        return (!($value & 1));
    } 
}

$isEvenFilter = new IsEvenFilter();

$isEvenFilter instanceof FilterFunction // makes sense
is_a($isEvenFilter, FilterFunction::class) // makes sense too
$isEvenFilter->__invoke(2) // makes sense too
```

If we apply that pattern to the most interesting interface with the longest name, and adding `Interface` (because of FIG by-laws) we get:
```php
interface ServerRequestHandlerFunctionInterface {}
```

Very long, but I believe it is appropriate. As the [Strategy pattern](https://en.wikipedia.org/wiki/Strategy_pattern) is about _interchangeable algorithms_, we could stretch that meaning a bit and use the suffix `Strategy` too:

```php
interface ServerRequestHandlerStrategyInterface {}
```

Fortunately, the namespace is already `Psr\Http\Message\Strategies`[1], and we can drop that Suffix. That would be similar to `Model\Employee[Model]` etc.:

```php
interface ServerRequestHandlerInterface {}
```

Thus, I believe the `*HandlerInterface` names (introduced with #4) are at least not so bad already – OFC, in contrast to the `*OperatorInterface` names.

[1]: The `Psr\Http\Message\Strategy` namespace would probably fit better.

**Comment by mindplay-dk** [comment]
> The more I think about it, the more I believe the most accurate name for `__invoke` interfaces include a `Function` suffix

From what I was taught about programming, interfaces ending in `Interface` are as wrong as functions ending in `Function`, variable-names ending in `_string` or `_int`, etc. - because all of those are polluting the application domain terminology with terms from the *language* domain.

Worse, what you're now describing as "functions" are actually supposed to be describing a facet, an area of responsibility, the role or capability of the class that is implementing it.

PSR interfaces ending in `Interface` is something we've had to reluctantly put up with, but this kind of practice has a very long bad standing in most other languages, including most other dynamic languages like JS, Ruby, etc.

The magic method-name `__invoke()` is a technical detail of the language, as indicated by the word "invoke", which belongs to the language-domain, and therefore should not occur in your domain.

If you would stick to ordinary, obvious, logical, everyday naming conventions for interfaces and method-names, the questions you should be asking are:

* For method-names: *what does the implementor do?*
* For interfaces-names: *what is it's role when it does what it does?*

For example, if you're defining the role of something that takes a Request and returns a modified Request, and you ask, what is it's role? I would say, it's a "request filter". And what does it do? It "filters the request".

Hence:

    interface RequestFilter {
        public function filterRequest(RequestInterface Request): RequestInterface
    }

It can't get anymore obvious and simple than that.

And yes, this allows an implementor to implement multiple facets - to have multiple roles in different contexts. That's what interfaces are for.

And yes, I know, we have to follow the backwards `Interface` suffix convention and carry on the legacy of this flawed decision, so:

    interface RequestFilterInterface {
        public function filterRequest(RequestInterface Request): RequestInterface
    }

Do you see how it already makes less sense just by the introduction of the word "interface", because the implementation doesn't by any means have the role of being an "interface" to anything? We're trying to describe the abstraction in terms of an object's role or responsibility in a domain context, but we're mixing in domain terminology from the programming-language.

I know we can't fight the `Interface` suffix convention because that would create inconsistency in the PSR space, but let's please not do anymore of this nonsense than we absolutely have to.

I do understand what it is you're trying to do, but PHP doesn't *have* typedefs, so you can't have what you want, not in PHP, not without stepping outside idiomatic uses of various language features; not without abusing various language concepts to some degree.

Besides, we have anonymous classes now, which, in those cases where you want something akin to a typedef, you can implement the relevant interface in-place using a throw-away instance of an anonymous class, e.g.:

```php
$app->addRequestFilter(
    new class implements RequestFilterInterface {
        public function filterRequest(RequestInterface Request): RequestInterface {
            // ...
        }
    }
);
```

It's the closest effective equivalent to a typedef, and no, it doesn't look sexy, yes, it's verbose, but it's *idiomatic* - this is the language feature you have for that purpose, or at least the closest best fit, and certainly cleaner and much more idiomatic than a bunch of mysterious `__invoke()` methods.

(oh, man, and I said I wouldn't get involved...)

**Comment by mindplay-dk** [comment]
PS: I don't know that it is, but, if support for callables is very important, you can include in the package a set of complementary wrapper-classes, e.g.:

```php
class RequestFilter implements RequestFilterInterface {
    public function __construct(callable $func) {
        // ...
    }
    // ...
}
```

Which makes a suitable alternative to anonymous classes:

```php
$app->addRequestFilter(new RequestFilter(function (RequestInterface $request) {
    // ...
}));
```

And yeah, that doesn't give you static type-checks, but that's a well-known consequence of using callables in a language without typedefs, and, if you care about type-checks, then don't use callables, use anonymous classes or actual classes instead, both of which are well-known idiomatic quantities.

I understand you're trying to make things more convenient, but you're actually making things less convenient, by adding learning-curve - when you diverge from idiomatic use of language-feature to "clever" or "creative" uses, you're forcing people to think about simple language features in a complex way; this perceived convenience, in practice causes complexity.

Drawing on the strengths of a language, in my point of view, also means recognizing it's weaknesses, by using the simplest language features applicable to the problem at hand - this is always going to be easier for members of the community of that language to understand, rather than attempting to emulate missing features from other languages.

And no, that doesn't always lead to perfect results - in particular, you can't expect, guarantee or *enforce* perfect static analysis in client-code, but that's the *client's* choice.

If you gave them the *option* to do things "the right way" (according to your ideas and mine) that's always going to be simpler, and overall better for everyone, than distorting the language idioms to try to force, lead or control what other developers do.

(yeah, this is me not getting involved, hmm.)

**Comment by mindplay-dk** [comment]
(I'm going to shut up now.)

**Comment by schnittstabil** [comment]
Aww, the `Function` suffix above should not have to be understood as some language internal thingy. (I&nbsp;would have chosen `Closure` if I've wanted to express that). It should be understood as other common suffixes which describe the role it plays. For example the `FilterFunction` above describes its role like `ContactView`, `ContactModel` and `ContactController` does, it describes what an object can do for me – `Function` does not necessarily imply an `__invoke` method, it might also provide a `call`, `apply` or `execute` method.

Should these roles _always_ be expressed by Suffixes? Well, no, but sometimes. IMO it should depend on the namespaces in place, like `Model\Contact` etc. Nevertheless, I agree, it's not a good rule of thumb to put `Function` everywhere.

---

But you put me back onto the right track. I love the correlation of OOP Design and natural language; especially the noun/verb metaphor, where _nouns_ are designed as _objects_ and _verbs_ as _methods_. As far as I can see, we agree on it's usefulness.

So, I try to describe my point of view using this metaphor focusing on the client side:

> ##### Why Bother?
> 1. The client creates a (plain) request object.
> 2. The client augments that request object.
> 3. The client sends the request object to the server and receives the produced response.
> 4. …

* The nouns are `client`, `server`, `request (object)` and `response`.
* The verbs are `create`, `augment`, `send` and `receive`.

This (almost) directly leads to a `Client` class and a _`Consumer`_ class, usually similar to:
```php
class Client
{
    protected function augmentRequest($request) {…}

    public function sendRequest($request) {
        $request = $this->augmentRequest($request);
        …
        return $response;
    }
}

class Consumer
{
    protected $client;

    public function __construct($client)
    {
        $this->client = $client;
    }

    protected function createPlainRequest() {…}

    public function foo()
    {
        $request = $this->createPlainRequest();
        $response = $this->client->sendRequest($request);
        …
    }
}
```

Obviously these have some drawbacks, notably they violate SOLID principles. The idea of this PSR is to standardize the [Strategy Pattern](https://en.wikipedia.org/wiki/Strategy_pattern), which is often used in that situation. Let's apply this pattern to the `Client` and `Consumer`:

![clientconsumerstrategies](https://cloud.githubusercontent.com/assets/6059032/23258739/d8545c3e-f9c9-11e6-9518-b1911e727959.png)


**Side note:** Obviously, PSR-17 provides us already an appropriate `CreatePlainRequest` interface.

The ![composition-arraow](https://cloud.githubusercontent.com/assets/6059032/23256321/3c6e15d0-f9bf-11e6-9920-46b982c83312.png)-arrow denotes composition, which is usually implemented like this:

```php
class Client
{
    protected $augmentRequest;
    protected $sendRequest;

    public function __construct(AugmentRequest $augmentRequest, SendRequest $sendRequest)
    {
        $this->augmentRequest = $augmentRequest;
        $this->sendRequest = $sendRequest;
    }

    public function sendRequest($request) {
        // call the $this->augmentRequest somehow:
        $request = $this…($request);
        // call the $this->sendRequest somehow:
        return $this…($request);
    }
}
```

The question which now arises is about the method names of `AugmentRequest` and `SendRequest`.

@mindplay-dk If I understand you correctly, then you say the right choice **for the method name**  would be:

```php
interface AugmentRequest
{
    public function augmentRequest($request);
}
```

Which means `Client::sendRequest` would look like:

```php
public function sendRequest($request) {
    $request = $this->augmentRequest->augmentRequest($request);
    return $this->sendRequest->sendRequest($request);
}
```

IMO, that would be anything but idiomatic. @mindplay-dk As far as I can see, you suggest to fix that by using a different interface name and/or a different property name. But if I look at the [examples at Wikipedia](https://en.wikipedia.org/wiki/Strategy_pattern), then I do not fully understand why you believe that would be strictly necessary.

It seems to me that, you derive an agent noun (e.g. `RequestSender`)  from a verb (`send` request), only because classes usually are named by nouns. But, in pure OOP languages everything is an object: integers, floats, characters, strings and even functions; the only exception I can think of at the moment are methods. As already stated at https://github.com/http-message-strategies-interop/fig-standards/issues/2#issuecomment-281617660, anonymous functions are objects in PHP too:

```php
$addOne = function ($n) {
    return $n + 1;
};

is_object($addOne) // => TRUE
```

IMO, this is not a technical detail, it's fully in line with OOP. And fixing the variable name (e.g. `$oneAdder`?) would not be idiomatic, is it? My point is, in OOP we usually use nouns for objects, but not always.

As we talk about _Strategies_, i.e. _interchangeable algorithms_, I believe the following method names would be much more appropriate, illustrated using the `Client:sendRequest` method:

```php
class Client
{
    //…
    public function sendRequest($request) {
        $request = $this->augmentRequest->execute($request);
        $request = $this->augmentRequest->call($request);
        $request = $this->augmentRequest->invoke($request);
        // …
    }
}
```

That said, I'm still in favor of `__invoke`, which in my opinion is more natural:

```php
class Client
{
    //…
    public function sendRequest($request) {
        // PHP >= 7
        $request = ($this->augmentRequest)($request);

       // or if you prefer this
       $request = call_user_func($this->augmentRequest, $request);
    }
}
```

And according to my (very detailed) explanation, I believe it is valid and accurate to have verbs as interface names – at least in this case.

**Comment by mindplay-dk** [comment]
> I believe it is valid and accurate to have verbs as interface names – at least in this case.

I disagree.

Interface names like `AugmentRequest` are wrong, and this particular case is particularly bad.

Flip the question on this type, and ask it: "what are you?" - the answer is, "I'm an augment request", because this is a type, so I'm expecting a noun, not a verb.

The type has to be a verb - the thing that an object becomes, or the role that an object takes on, by implementing that interface. It should describe a role or responsibility in some context.

You're still trying to treat interfaces as a kind of type-definition for a function.

In PHP, there is no such thing.

> That said, I'm still in favor of `__invoke`, which in my opinion is more natural

You can almost tell from the example you gave just how unnatural this is:

    $request = ($this->augmentRequest)($request);

This looks bizarre.

The only time you should use `__invoke()` is in place of something that expects a callable.

What you're trying to do is the opposite: require it to be a callable, but a specific implementation of a callable, which would be a typedef, which isn't a thing in PHP.

    $request = $this->request_filter->filterRequest($request);

Verbose? Yes. Would I like shorter syntax? Hell yes, but PHP doesn't have that, and I'm not going to abuse the language in favor of cosmetics, I'm going to use the language features idiomatically, as intended.

And I don't think you addressed the issue of being unable to implement more than one of these interfaces. What if my component is a request filter as well as a request handler? Or any other two roles in different contexts. Now I need adapters or something, for every class, and factory-methods for those adapters, to be able to pass my object to type-hinted methods that consume my component in one of it's roles.

I say again, this is *not* what interfaces are for.

Please stop trying to outsmart the language. The language is flawed. Deal with it :-)

**Comment by schnittstabil** [comment]
> You're still trying to treat interfaces as a kind of type-definition for a function.

Nope, and I do not think `= ($this->augmentRequest)($request)` is smart or sexy – I hate the additional parentheses. But that's the syntax of PHP, and a way to deal with bad design decisions made many, many years ago. If we remove the first pair of parentheses, then it perfectly fits.

I do not want to outsmart the language; and I do not think it is flawed – it is just not _feature complete_. Moreover, I want to use all features the language, not PHP 4 only :wink: Would typedefs solve this issue too; or better? – In my opinion, debates about that take us nowhere: we do not have them and PHP may provide a totally different solution at some day.

> Interface names like `AugmentRequest` are wrong, and this particular case is particularly bad.
> Flip the question on this type, and ask it: "what are you?" - the answer is, "I'm an augment request", because this is a type, so I'm expecting a noun, not a verb.

Well, that issue can be easily explained – you've forgotten the namespace:
*  "`Acme\View\Contact` what are you?": "I'm a contact." – Err, no, "I'm a contact **view**."
*  "`…\Strategy\AugmentRequest` what are you?": "I'm an augment request **strategy**."

>  the questions you should be asking are:
> * For method-names: *what does the implementor do?*
> * […]

I've done exactly that: the implementor of an _augment request **strategy** applies its strategy to a `$request`_
– Hence, `…\Strategy\AugmentRequest::apply($request)` is idiomatic, isn't it?

> And I don't think you addressed the issue of being unable to implement more than one of these interfaces.

I do not think that would be helpful at this moment – currently I want good names only. But, you've asked; and obviously do not know the last [http-middleware/issues](https://github.com/http-interop/http-middleware/issues).
Many interface can be implemented by a single class in PHP 5.6 (see an [DelegateAndMiddleware](https://github.com/schnittstabil/symfony-httpinterop-bridge/blob/00a8fff1d369afed22a4718cc2de47f71134ae61/src/HttpInteropToSymfonyBridge.php#L23) for example), and in PHP 7.2 a single class can implement all interfaces (see [PHP RFC: Parameter Type Widening](https://wiki.php.net/rfc/parameter-no-type-variance)). Should they be allowed to? Hell no! That would violate LSP, because the phpdoc includes `@return RequestInterface` or `@return ResponsInterface` (return type declarations would successfully prevent that). 

Long story short:
1. PHP 5.6 already allows to implement some of the interfaces in a single class.
2. PHP 7.2 allows to implement all interfaces in a single class, for example `RequestHandler` and `ServerRequestHandler` – which is bad.
3. Using return type declarations, LSP violations in 2. can be prevented in PHP 7.2.

In your question, you've chosen `RequestFilter` and `RequestHandler`. Using return type declarations would indeed prevent that; but you know, I'm very creative with use case examples. I have many, many use cases for these interfaces, but not one where combining those two – and I doubt that there are valid real-world use cases, because they differ in return types. – If you know an example, please share it.

**Comment by mindplay-dk** [comment]
You've decided that implementing more than one interface is somehow "bad" - I think that's nonsense; interfaces exist for precisely that reason, so that the same object can serve multiple roles in different contexts. That doesn't violate any principles.

But you've decided you don't think there's a use-case for implementing more than one of these interfaces in the same class, and you don't think people should be allowed to do that.

I think you're wrong.

I don't understand why you can't just do the idiomatic, obvious, simple thing, instead of delivering these mile-long justifications in favor of something that a lot of developers are likely to find odd, annoying or confusing, and which *may* obstruct them from doing really obvious things the obvious way.

I'm sorry, but I can't see forcing use of `__invoke()` on anybody as anything other than dictating architecture. It's opinionated. It's totally unnecessary, and there's an obvious, clean, simple, idiomatic alternative - and therefore, I don't understand where it is you're coming from, what it is you're trying to accomplish, or why you would make such choices.

So I am once again opting out of this debate, sorry - I just don't have the energy.

**Comment by schnittstabil** [comment]
> You've decided that implementing more than one interface is somehow "bad"
> […]
> But you've decided you don't think there's a use-case for implementing more than one of these interfaces in the same class, and you don't think people should be allowed to do that.

Sorry, but why do you believe all that? Was my last comment so confusing? None of that is true.

If you can provide me one single use case, which is prevented by these interfaces and cannot be solved in a much, much better way, then I will immediately close this issue and propose non-invoke interfaces.

> I don't understand why you can't just do the idiomatic, obvious, simple thing

Sorry, but to me, the first two versions below are idiomatic, obvious and simple. And the third version, which I get when I use your approach, appears constructed, opinionated and totally unnecessary.

```php
class Client
{
    //…
    public function sendRequest($request) {
        $request = $this->augmentRequestStrategy->apply($request);
        $request = ($this->augmentRequest)($request);

        $request = $this->requestAugmenter->augmentRequest($request);

        // …
    }
}
```

Or, did I misunderstood your approach? What did I do wrong in opinion?

**Comment by mindplay-dk** [comment]
I'm sorry, now you do want to use real method-names?

I'm a bit under the weather, I have a cold for like a week now, maybe I'm not keeping up with all your comments... Sorry about that :-/

Regarding namespace/class-names.

> "`Acme\View\Contact` what are you?": "I'm a contact." – Err, no, "I'm a contact view."

According to it's class-name, this class is a `Contact` model which was put in the wrong namespace.

With proper imports, all you'll see in client code is e.g.:

    $contact = new Contact();

Near the top of the file, of course you'll see a use-statement, but nobody writes out fully-qualified class-names throughout the entire file, so this gets messy.

I'd have to mentally "import" fully-qualified class-names into my memory, and then "parse" the code.

I'd also very likely need to qualify `Contact` from `Contact` somehow, in a controller that depends both on `Model\Contact` and `View\Contact`, I'd likely need to rename one of them locally, which gets even more confusing and harder to parse.

In my opinion, namespaces are a means of grouping classes - I don't think of namespaces as being part of the class-names.

Besides, views usually have a more specific purpose, and there are often multiple views relating the same entity, so likely I would name my views something like `ContactEditForm`, `ContactView`, etc.

In my opinion, down the line, you will regret naming your view-models as though they were entities.

**Comment by schnittstabil** [comment]
> I'm sorry, now you do want to use real method-names?

I'm still open to this. But, `RequestAugmenter::augmentRequest` makes not much sense to me, it is not _DRY_ in the linguistic sense.

> In my opinion, namespaces are a means of grouping classes - I don't think of namespaces as being part of the class-names.

Granted, that makes sense. The main problem I wanted to solve are the lengths of the interface names, they are already very long, e.g.:

```php 
namespace Psr\Http\Message\Strategy;

interface HandleServerRequestStrategyInterface {…}
// vs
interface HandleServerRequestInterface {…}
```

**Comment by schnittstabil** [comment]
> I'm sorry, now you do want to use real method-names?

Oh, sorry, this needs some clarification: I will immediately go with real method-names, **if** one can provide me a single (almost reasonable) use-case, which is prevented by these `__invoke`-interfaces. This means, it needs some small **code** snippet to convince me.

**Comment by mindplay-dk** [comment]
> `RequestAugmenter::augmentRequest` makes not much sense to me, it is not DRY in the linguistic sense.

The DRY principle does not apply to names.

You object to this because you don't like having the same language in the interface definitions more than once? That also explains why you want to remove important information from class-names that occurs in namespace-names then, I take it.

I think that's misguided. Method-names, namespaces and class-names are three different things, they have different names, for different reasons - if the same language occurs in those names, that's just simply each of those names being descriptive enough to function on their own.

You're not "repeating" anything - class names and (especially implementations of interface) method-names should make sense on their own.

You're not helping anybody, or "optimizing" in any sense, or making things in any way cleaner or clearer, by eliminating important domain language from class-names or method-names - you're only making things ambiguous and muddy.

> This means, it needs some small code snippet to convince me.

Why do you require evidence or proof to convince you to do the logical, obvious, idiomatic thing?

The *opposite* should require proof.

In my opinion, the burden should be on you to demonstrate why or how forcing use of `__invoke()`, and thereby dictating architecture (by removing the freedom to implement multiple interfaces) is more than opinionated.

Show me a use-case that *enables* you to do something useful or unique, or show me why or how `__invoke()` in interfaces is useful or necessary.

Don't try to *justify* it - if it's really useful or necessary, it should demonstrate it's own usefulness *without* the complex, lengthy ratifications.

I guess, really, what I'm saying is, I don't think this *is* useful or necessary, and I think you've already demonstrated that it's not. You seem to be attached to this idea for subjective reasons - because you, personally, don't think someone should implement more than one interface at a time, but I can't view this as anything other than opinion, and I don't think believe it's necessary or productive to diverge from simple, idiomatic, everyday code to try to enforce certain opinions.

Interfaces will have more universal applicability, and provide overall better interoperability across systems, if you make things simple and un-opinionated.

If you need me to cite principles, that's KISS, plain and simple: don't introduce complexity, or artifacts, or impose arbitrary constraints or limitations on client code.

**Comment by schnittstabil** [comment]
> The DRY principle does not apply to names.

I'm fully aware of that, but I've written "_DRY_ in the linguistic sense". The word "DRY" was set in italics for reasons.

>> This means, it needs some small code snippet to convince me.

> Why do you require evidence or proof to convince you to do the logical, obvious, idiomatic thing?

At first you've taken that out of context. It has almost nothing to do with `__invoke` vs _real_ method names. It refers to your **name conflict** concerns:
1. If someone provide me a single (almost reasonable) use-case, which is prevented by the name conflict, then:
2. I will immediately go with **different method names** to solve that issue, which in effect means:
3. I will immediately go with **real method names**

Secondly, it seems you believe I try to justify something, which is clearly not "logical, obvious, idiomatic".
We obviously have different backgrounds and we both got used to different coding styles. I can assure you, an `__invoke` interface and its usage is idiomatic to me. – You may take a look at some examples at github, I believe I'm not alone: [`Strategy and __invoke search results`](https://github.com/search?l=PHP&q=strategy+__invoke&type=Code&utf8=%E2%9C%93).

---

> Show me a use-case that enables you to do something useful or unique, or show me why or how `__invoke()` in interfaces is useful or necessary.

You can already find some examples at [github.com/http-message-strategies-interop](https://github.com/http-message-strategies-interop):

* [example-middleman](https://github.com/http-message-strategies-interop/example-middleman)
* [example-slim](https://github.com/http-message-strategies-interop/example-slim)
* [example-zend-stratigility](https://github.com/http-message-strategies-interop/example-zend-stratigility)
* and many more are planned

---

> you, personally, don't think someone should implement more than one interface at a time

No, I do not think that, and the following is a very reasonable example, which works in PHP 7.2:

```php
interface RequestHandlerInterface
{
    public function __invoke(RequestInterface $request): ResponseInterface;
}

interface ServerRequestHandlerInterface
{
    public function __invoke(ServerRequestInterface $request): ResponseInterface;
}

class Handler implements RequestHandlerInterface, ServerRequestHandlerInterface
{
    public function __invoke($request): ResponseInterface
    {
        if (!($request instanceOf RequestInterface)) {
            throw new InvalidArgumentException(…);
        }

        return (new Response())->withAttribute('X-PoweredBy', 'Unicorns');
    }
}
```

See, [PHP RFC: Parameter Type Widening](https://wiki.php.net/rfc/parameter-no-type-variance) for details – and as far as I can see the `if (…) {}` won't be necessary too, sooner or later.

With different method names, we MUST do:

```php
class Handler implements RequestHandlerInterface, ServerRequestHandlerInterface
{
    public function handleRequest(RequestInterface $request): ResponseInterface
    {…}

    public function handleServerRequest(ServerRequestInterface $request): ResponseInterface
    {…}
}
```
The problem here is not on the interface implementer side. The problem is that, the caller must decide which method to call! That would be really bad, wouldn't it? For that reason, I want to have a _name conflict_ at least for those two interfaces – but this has nothing to do with `__invoke`.

---

@mindplay-dk I believe, it became hard to track all the different versions and their state, and we start to go around in circles. I try to find some time to create multiple alternative PRs this week, which incorporate the current state of debate.

**Comment by mindplay-dk** [comment]
> The problem is that, the caller must decide which method to call!

Well, of course?

Just as it must decide which interface to type-hint against in methods.

If I'm implementing an interface and method, of course I must decide whether I can handle all `RequestInterface` or only specifically `ServerRequestInterface`.

How is that bad?

You picked a very odd example - those two particular interfaces, `RequestHandlerInterface` and `ServerRequestHandlerInterface`, are essentially mutually exclusive; there would be no reason to implement both, because the two contracts state, you either handle *all* requests, or you handle *server*-requests *only*.

There would be no sense in implementing both, and conversely, someone implementing a component that accepts a request-handler would need to decide if that component supports request-handlers that accept *all* requests, or request-handlers that accept *server*-requests *only*.

**Comment by schnittstabil** [comment]
At first, it would be bad, because it wouldn't be idiomatic.

Similar to the reason, why PSR interfaces ending in `Interface` are bad: `handleServerRequest` includes the type it's able to process. We got used to this pattern only, because we have had to deal with the badness of PHP's type-system:

```php
$requestHandler->handleServerRequest($serverRequest);
// => Hey $requestHandler, please handleServerRequest this $serverRequest

// vs:
$requestHandler->handle($serverRequest);
// => Hey $requestHandler, please handle this $serverRequest
```

> because the two contracts state, you either handle _all_ requests, or you handle _server_-requests _only_.

No, because every _server_-request is a request. And, for that reason every _RequestHandler_ can be seen as a _ServerRequestHandler_ too:

![contravarianceofhandlers](https://cloud.githubusercontent.com/assets/6059032/23402695/fac36f6c-fdac-11e6-9e19-3b1d555a59cf.png)

That is called [contravariance of the method argument type](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)#Contravariant_method_argument_type); and [Parameter Type Widening](https://wiki.php.net/rfc/parameter-no-type-variance) is the first step in this direction.

[http-message-strategies-interop/example-guzzle-get](https://github.com/http-message-strategies-interop/example-guzzle-get) is an example which is able to process every `RequestInterface` and therefore every `ServerRequestInterface` as well:

```php
use GuzzleHttp\Psr7\Request;
use Zend\Diactoros\ServerRequest;

$gget = new GuzzleGet();

$response0 = $gget(new Request('GET', 'https://httpbin.org'));
$response1 = $gget(new ServerRequest([], [], 'http://httpbin.org', 'GET'));
```

If you need use-cases on the sever-side, then think about load-balancing, proxies etc.

Actually `GuzzleGet` does not implement `ServerRequestHandlerInterface`, it already works as expected – already today, in PHP 5. Thus, `$gget` also implements the `ServerRequestHandlerInterface` contract; and because of that, in many languages `$gget instanceof ServerRequestHandlerInterface` automatically  returns `true`. Sadly, PHP is an exception and `GuzzleGet` must explicitly implement `ServerRequestHandlerInterface` to work around the badness of the type-system.

Secondly, it would be bad for implementers and consumers.

```php
class Handler implements ServerRequestHandlerInterface, RequestHandlerInterface
{
    public function handleRequest(RequestInterface $request)
    {…}

    public function handleServerRequest(ServerRequestInterface $request)
    {
        return $this->handleRequest($request);
    }
}
```

The problems I see here:
1. if `$request instanceof ServerRequestInterface`, then `handleRequest` must provide the same semantics as `handleServerRequest`. Otherwise, IMO it would not act as consumers expect.
2. If we extend the `ServerRequestInterface`, then we do not really know which method to call, e.g. in the most worst situation: `interface StupidMessageInterface extends RequestInterface, ServerRequestInterface {}` – yes, I know this is a very minor issue in our domain.
3. As `ServerRequestInterface` extends `RequestInterface`, the handler should know better than the caller how to handle different `$request` types in my opinion. Why should we burden the caller with this decision? KISS applies here, doesn't it? :wink:

**Comment by mindplay-dk** [comment]
> Similar to the reason, why PSR interfaces ending in `Interface` are bad: `handleServerRequest` includes the type it's able to process

That is not the same thing at all - interfaces ending in `Interfaces` are not bad because they include redundant information, but because they include terminology from the *language* domain, as opposed to `handleServerRequest` which is domain-language entirely.

Again, DRY does not apply to class-names or function-names - not in any sense.

In fact, it's almost the opposite - method-names *should* make sense in isolation, because they will appear only *once* in the context of the class that implements the method, while it will occur *many* time in call-sites in different contexts; the name should explain clearly what it's doing, even when taken out of declaration context, which is not where you will see it most of the time.

> We got used to this pattern only, because we have had to deal with the badness of PHP's type-system

No, we do it to prevent collisions.

Method-names like `process()` are more likely to collide with other (equally poorly named) methods, and thereby preventing implementation of multiple interfaces - names containing more domain terminology, such as for example `processRequest()` and `processPayment()`, are less likely to collide, and are more easily understood at call-sites.

You're right that this is partially because of the language, but features like [explicit interface implementation](https://msdn.microsoft.com/en-us/library/ms173157.aspx) only really works in static languages; it wouldn't work in PHP, and since we can't *call* or *declare* interface methods explicitly, it's extra important we qualify operations by *naming* the methods explicitly.

> > because the two contracts state, you either handle all requests, or you handle server-requests only.

> No, because every server-request is a request. And, for that reason every RequestHandler can be seen as a ServerRequestHandler too:

That is precisely what I was saying - a request-handler handles *all* requests, and a  server-request-handler handles server-requests only.

I don't know how we manage to miscommunicate this badly all the time - as I've said, I honestly don't think the written medium is suitable for these discussions, at all, and I'd really rather not participate.

This is exhausting and I really don't have time for it, sorry.

**Comment by schnittstabil** [comment]
> I don't know how we manage to miscommunicate this badly all the time

I do not understand that either.

You've said:
> `RequestHandlerInterface` and `ServerRequestHandlerInterface`, are **essentially mutually exclusive**; there would be **no reason to implement both**

I've said, every `RequestHandlerInterface` instance already implements the `ServerRequestHandlerInterface` contract, thus they cannot be *mutually exclusive* and the reason to explicit implement both is because of the PHP type-system.

You think `handle()` is too generic, because you can think of situations where one object wants to provide `handleRequest()` and `handlePayment()`. Fair enough, but that is a new argument, taking that into account, I end up with:
![handlerinheritance](https://cloud.githubusercontent.com/assets/6059032/23450534/ffc8c552-fe5b-11e6-82b2-ab28b52406da.png)

The reasons for not having `ServerRequestHandler::handleServerRequest`:
1. `RequestHandlerInterface::handleRequest` naturally overrides `ServerRequestHandler::handleRequest`.
2. `ServerRequestHandler::handleServerRequest` leads to ambiguities: the caller does not know which method to call.
3. `ServerRequestHandler::handleServerRequest` complicates caller code, e.g:
```php
if ($handler instanceof ServerRequestHandlerInterface) {
    $response =  $handler->handleServerRequest($request);
else if ($handler instanceof RequestHandlerInterface) {
    $response =  $handler->handleRequest($request);
} else {
    throw new IllegalArgumentException(…);
}
```
4. the caller code usually must be refactored if he wants to support `AwesomeNewRequestInterface implements RequestInterface` too, e.g. `$handler->handleAwesomeNewRequest($request)`. And the same holds for handlers.

Why do you believe that would be desirable? My argument is not about redundant information/DRY, my point is that `handleServerRequest` includes the type `ServerRequest` for no reason.

**Comment by mindplay-dk** [comment]
Okay, I agree with points 1 through 3, and I'm fine with `handleRequest()` as the method-name in both interfaces - the key for me is having the word "request", which I think is enough domain language to prevent collisions, and to make sense of these methods when taken out of context.

I'm not sure I understand point 4, but it sounds like we're finally on the same page now anyway :-)

**Comment by mindplay-dk** [comment]
Well, the whole discussion about the handler interface details is now sunk cost, because PSR-15 will standardize this interface.

All that remains then is the request->request and response->response interfaces.

I can't read through all of this again and don't recall where exactly we ended up - as I recall, you weren't too fond of the term "filter", and I'm not fond of "operator", what do you think about "transform"?

So like:

```php
interface RequestTransformInterface {
    public function transformRequest(ServerRequestInterface $request): ServerRequestInterface;
}

interface ResponseTransformInterface {
    public function transformResponse(ResponseInterface $request): ResponseInterface;
}
```

I hate the interface suffix, but that's the PSR way.

I really do think we should use real return type-hints - because why would we design new interfaces for an older version of PHP? Let's push for the modern approach.
